### ISSUE TITLE
RFC: compatibility with ipfs-companion

### ISSUE THREAD

**Issue by pgte** [ipfs-shipyard/peer-star-app#197]
Currently, peer-star-app runs an IPFS node for every tab for every domain. That's not optimal.
There have been some successful experiences of running peer-star-app in a browser extension (see the [Discussify browser extension](https://github.com/ipfs-shipyard/discussify-browser-extension)), but it's still a new installation away from being useful.

[ipfs-companion](https://github.com/ipfs-shipyard/ipfs-companion) is a browser extension that already exists and is developed and supported by Protocol Labs. When it's installed, it provides apps with an IPFS proxy object that talks to the IPFS instance running inside the browser extension (either js-ipfs or go-ipfs). It would be great if, instead of creating the app's IPFS node, that peer-star-app would use that IPFS node instead.

## Challenges

Currently, peer-star-app extends IPFS, implementing a custom sync protocol for ∂-CRDTs. It also makes use of the libp2p interface to wrap the websocket-star transport to provide a filter on top of discovery.

The reason peer-star-app does these things are the following:

- [ ] js-ipfs wants to connect to as many nodes as possible. Peer-star-app wants (small) group collaborations, but maintaining app-wide gossip in a scalable way.
- [ ] js-ipfs bitswap is not designed for syncing a DAG of operations (deltas). A peer-star-app replica wants to only send the required deltas another replica requires and do it in one burst.

Unless these problems are addressed in js-ipfs, peer-star-app still requires these things to be possible:

- [ ] Wrap the transport
- [ ] Filter discovery events
- [ ] Implement a custom delta sync protocol

/cc @ipfs-shipyard/experimentalists @lidel @olizilla @alanshaw

**Comment by lidel** [comment]
@pgte I imagine pubsub is a prerequisite for what you do, correct?
If so, then using IPFS Companion may be blocked by https://github.com/ipfs/js-ipfs-http-client/issues/518

Otherwise, something you can experiment with today is [window.ipfs-fallback](https://github.com/ipfs-shipyard/window.ipfs-fallback) library. It provides `getIpfs()` which will use `window.ipfs` if Companion is present and fallback to spawning js-ipfs node otherwise and aims to be future-proof, so is safer than using `window.ipfs` directly.

FYSA there is also a deeper discussion about the future of `window.ipfs` in https://github.com/ipfs-shipyard/ipfs-companion/issues/589, but as long you use fallback lib details should not concern you at this stage.

**Comment by pgte** [comment]
@lidel yes, PubSub is also a requisite, I didn't realise it wasn't supported. I'll add this to the list of requirements.

**Comment by lidel** [comment]
@pgte Yes, its probably too soon, just putting it on your radar if such need arises :)

Digression: In case you want to experiment with ipfs-companion before  ipfs/js-ipfs-http-client#518 is solved, you can take advantage of the fact it [supports two backends](https://github.com/ipfs-shipyard/ipfs-companion/blob/master/docs/node-types.md): go-ipfs over js-ipfs-http-client (stable default) or embedded js-ipfs (experimental). If you switch to experimental js-ipfs backend, then you will be able to specify custom configuration for that node (enable pubsub, set custom signaling server etc).

**Comment by pgte** [comment]
@lidel I would be ok with initially only supporting Companion in the "js-ipfs-only" mode. I've opened an issue on Companion to discuss this in more depth: https://github.com/ipfs-shipyard/ipfs-companion/issues/634
